tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,279 @@
+# SUSE's openQA tests
+ #
+ # Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+ 
+# Summary: A test on KVM host to pass a SR-IOV NIC VF to guest via libvirt.
+# Maintainer: Julie CAO <JCao@suse.com>
+ 
+use base "virt_feature_test_base";
+use strict;
+use warnings;
+use testapi;
+use set_config_as_glue;
+use utils;
+use virt_utils;
+use xen;
+use virt_autotest::virtual_network_utils qw(save_guest_ip test_network_interface);
+ 
+sub run_test {
```

Comment:
Suggest to extend a kind of test that attach and detach more than 1 vf sequentially, eg 2 or 3 vfs.

Reply:
It will make the test a bit more complex. the additional work I can think of is:
*1: one or more vf variables, and double attaching/detaching steps. it is not difficult to carry out by extending existing variables and reusing present functions.

*2: no more availale pci slot for the current guest. it allows only one pci device passthrough to guest, but it may be impleted by adding one more controller? I am not sure.

Reply:
> It will make the test a bit more complex. the additional work I can think of is:
> *1: one or more vf variables, and double attaching/detaching steps. it is not difficult to carry out by extending existing variables and reusing present functions.
> 

IMHO, it is valuable investment. Take your time! I believe  good way to support it can be figured out.

> *2: no more availale pci slot for the current guest. it allows only one pci device passthrough to guest, but it may be impleted by adding one more controller? I am not sure.

If I remember correctly, sle12sp5 have multiple pci ports for multiple vf passthru. For sle15sp2, Jim mentioned in one of leon's bug that pci controller configuration changed. So different products are a little different. And as you figured, we can change the guest setting to add 1 more pci controller with multiple ports.

Reply:
Working on it and adding Xen support now.

Reply:
Just found a capable way to support plugging more PCIe devices to guests. Working on extend the test scenarios now.